File: creational-design-pattern/prototype-pattern/prototype.py
"""
Prototype Pattern
=================
Category: Creational Design Pattern

Intent:
    Create new objects by cloning an existing instance (the prototype) rather
    than building from scratch. Useful when object creation is expensive or
    when you want to produce copies with slight variations.

When to use:
    - When creating an object is expensive (DB queries, file I/O, complex computation)
    - When you need many similar objects with small differences
    - When you want to avoid a proliferation of factory subclasses
    - When object configuration is complex and you'd rather copy-and-tweak

Key Participants:
    - Prototype (interface): Declares the clone method
    - ConcretePrototype: Implements clone by copying its own state
    - Client: Creates new objects by asking a prototype to clone itself

Deep vs Shallow Copy:
    - Shallow copy: copies primitive fields, but nested objects are shared references
    - Deep copy: recursively copies everything — changes to the clone don't affect original
    Always use deep copy unless you explicitly want shared nested state.
"""

import copy
from dataclasses import dataclass, field


# ---------------------------------------------------------------------------
# Prototype — Document Template System
# ---------------------------------------------------------------------------
@dataclass
class DocumentStyle:
    """Nested object to demonstrate deep vs shallow copy."""
    font: str = "Arial"
    font_size: int = 12
    color: str = "#000000"
    line_spacing: float = 1.5


@dataclass
class DocumentTemplate:
    """
    A document prototype that can be cloned.
    Deep copying ensures nested objects (like style) are independent copies.
    """
    title: str
    style: DocumentStyle
    sections: list[str] = field(default_factory=list)
    metadata: dict[str, str] = field(default_factory=dict)

    def clone(self) -> "DocumentTemplate":
        """
        Deep clone: creates a fully independent copy.
        Modifying the clone's style, sections, or metadata won't affect the original.
        """
        return copy.deepcopy(self)

    def shallow_clone(self) -> "DocumentTemplate":
        """
        Shallow clone: nested objects (style, sections, metadata) are SHARED.
        Mutating them on the clone WILL affect the original. Use with caution.
        """
        return copy.copy(self)

    def __str__(self) -> str:
        sections = ", ".join(self.sections) if self.sections else "(none)"
        meta = ", ".join(f"{k}={v}" for k, v in self.metadata.items()) or "(none)"
        return (
            f"Document: '{self.title}'\n"
            f"  Style: {self.style.font} {self.style.font_size}pt {self.style.color}\n"
            f"  Sections: {sections}\n"
            f"  Metadata: {meta}"
        )


# ---------------------------------------------------------------------------
# Prototype Registry — stores named prototypes for easy access
# ---------------------------------------------------------------------------
class TemplateRegistry:
    """
    A registry of pre-configured prototypes. Instead of constructing
    complex templates from scratch, clone a registered prototype and customize.
    """

    def __init__(self) -> None:
        self._templates: dict[str, DocumentTemplate] = {}

    def register(self, name: str, template: DocumentTemplate) -> None:
        self._templates[name] = template

    def create(self, name: str) -> DocumentTemplate:
        """Clone a registered prototype. Raises KeyError if not found."""
        prototype = self._templates.get(name)
        if prototype is None:
            raise KeyError(f"No template registered as '{name}'")
        return prototype.clone()


# ---------------------------------------------------------------------------
# Usage Example
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    # Set up a registry with pre-configured prototypes
    registry = TemplateRegistry()

    # Register a "report" prototype — expensive to configure from scratch
    report_style = DocumentStyle(font="Times New Roman", font_size=11, color="#333333")
    report_template = DocumentTemplate(
        title="Quarterly Report",
        style=report_style,
        sections=["Executive Summary", "Financials", "Outlook"],
        metadata={"author": "Template", "department": "Finance"},
    )
    registry.register("report", report_template)

    # Register a "memo" prototype
    memo_style = DocumentStyle(font="Helvetica", font_size=10, color="#000000", line_spacing=1.15)
    memo_template = DocumentTemplate(
        title="Internal Memo",
        style=memo_style,
        sections=["Subject", "Body", "Action Items"],
        metadata={"classification": "internal"},
    )
    registry.register("memo", memo_template)

    # Clone and customize — much cheaper than building from scratch
    q1_report = registry.create("report")
    q1_report.title = "Q1 2025 Report"
    q1_report.metadata["author"] = "Alice"
    q1_report.sections.append("Risk Assessment")

    q2_report = registry.create("report")
    q2_report.title = "Q2 2025 Report"
    q2_report.metadata["author"] = "Bob"

    team_memo = registry.create("memo")
    team_memo.title = "Engineering All-Hands Notes"
    team_memo.metadata["author"] = "Charlie"

    print("=== Original template (unchanged because we deep-copied) ===")
    print(report_template)
    print()
    print("=== Q1 Report (cloned + customized) ===")
    print(q1_report)
    print()
    print("=== Q2 Report (cloned + customized) ===")
    print(q2_report)
    print()
    print("=== Team Memo (cloned from memo template) ===")
    print(team_memo)

    # Demonstrate deep vs shallow copy difference
    print("\n=== Deep vs Shallow Copy Demo ===")
    deep = report_template.clone()
    shallow = report_template.shallow_clone()

    deep.style.color = "#FF0000"     # Only affects the deep clone
    shallow.style.font_size = 99     # ALSO affects the original!

    print(f"Original font_size after shallow clone mutation: {report_template.style.font_size}")
    print(f"Original color after deep clone mutation: {report_template.style.color}")

File: creational-design-pattern/prototype-pattern/prototype.rs
//! Prototype Pattern
//! =================
//! Category: Creational Design Pattern
//!
//! Intent:
//!   Create new objects by cloning an existing instance rather than building
//!   from scratch. Useful when object creation is expensive or when you want
//!   copies with slight variations.
//!
//! When to use:
//!   - When creating an object is expensive (DB queries, complex computation)
//!   - When you need many similar objects with small differences
//!   - When object configuration is complex and you'd rather copy-and-tweak
//!
//! Key Participants:
//!   - Prototype: Implements Clone trait
//!   - Client: Creates new objects by calling .clone() on prototypes
//!
//! Rust approach:
//!   Rust has the Clone trait built into the language. Deriving Clone provides
//!   deep copy semantics for all fields that themselves implement Clone.
//!   This makes the Prototype pattern almost trivial in Rust.

use std::collections::HashMap;

// ---------------------------------------------------------------------------
// Product types — derive Clone for deep copy support
// ---------------------------------------------------------------------------

/// Nested struct to demonstrate that Clone is deep (not shared references).
#[derive(Clone, Debug)]
struct DocumentStyle {
    font: String,
    font_size: u32,
    color: String,
    line_spacing: f64,
}

/// The prototype: a document template that can be cloned and customized.
/// Deriving Clone gives us deep copy of all fields including nested structs,
/// Vec, and HashMap.
#[derive(Clone, Debug)]
struct DocumentTemplate {
    title: String,
    style: DocumentStyle,
    sections: Vec<String>,
    metadata: HashMap<String, String>,
}

impl DocumentTemplate {
    fn new(
        title: &str,
        style: DocumentStyle,
        sections: Vec<&str>,
        metadata: Vec<(&str, &str)>,
    ) -> Self {
        Self {
            title: title.to_string(),
            style,
            sections: sections.into_iter().map(String::from).collect(),
            metadata: metadata
                .into_iter()
                .map(|(k, v)| (k.to_string(), v.to_string()))
                .collect(),
        }
    }
}

impl std::fmt::Display for DocumentTemplate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let sections = if self.sections.is_empty() {
            "(none)".to_string()
        } else {
            self.sections.join(", ")
        };
        let meta = if self.metadata.is_empty() {
            "(none)".to_string()
        } else {
            self.metadata
                .iter()
                .map(|(k, v)| format!("{}={}", k, v))
                .collect::<Vec<_>>()
                .join(", ")
        };
        write!(
            f,
            "Document: '{}'\n  Style: {} {}pt {}\n  Sections: {}\n  Metadata: {}",
            self.title, self.style.font, self.style.font_size, self.style.color, sections, meta
        )
    }
}

// ---------------------------------------------------------------------------
// Prototype Registry — stores named prototypes
// ---------------------------------------------------------------------------
struct TemplateRegistry {
    templates: HashMap<String, DocumentTemplate>,
}

impl TemplateRegistry {
    fn new() -> Self {
        Self {
            templates: HashMap::new(),
        }
    }

    fn register(&mut self, name: &str, template: DocumentTemplate) {
        self.templates.insert(name.to_string(), template);
    }

    /// Clone a registered prototype. Returns None if not found.
    fn create(&self, name: &str) -> Option<DocumentTemplate> {
        self.templates.get(name).cloned() // .cloned() calls Clone on the value
    }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
fn main() {
    let mut registry = TemplateRegistry::new();

    // Register pre-configured prototypes
    registry.register(
        "report",
        DocumentTemplate::new(
            "Quarterly Report",
            DocumentStyle {
                font: "Times New Roman".to_string(),
                font_size: 11,
                color: "#333333".to_string(),
                line_spacing: 1.5,
            },
            vec!["Executive Summary", "Financials", "Outlook"],
            vec![("author", "Template"), ("department", "Finance")],
        ),
    );

    registry.register(
        "memo",
        DocumentTemplate::new(
            "Internal Memo",
            DocumentStyle {
                font: "Helvetica".to_string(),
                font_size: 10,
                color: "#000000".to_string(),
                line_spacing: 1.15,
            },
            vec!["Subject", "Body", "Action Items"],
            vec![("classification", "internal")],
        ),
    );

    // Clone and customize
    let mut q1_report = registry.create("report").expect("report template not found");
    q1_report.title = "Q1 2025 Report".to_string();
    q1_report.metadata.insert("author".to_string(), "Alice".to_string());
    q1_report.sections.push("Risk Assessment".to_string());

    let mut q2_report = registry.create("report").expect("report template not found");
    q2_report.title = "Q2 2025 Report".to_string();
    q2_report.metadata.insert("author".to_string(), "Bob".to_string());

    let mut team_memo = registry.create("memo").expect("memo template not found");
    team_memo.title = "Engineering All-Hands Notes".to_string();
    team_memo.metadata.insert("author".to_string(), "Charlie".to_string());

    println!("=== Q1 Report (cloned + customized) ===");
    println!("{}\n", q1_report);
    println!("=== Q2 Report (cloned + customized) ===");
    println!("{}\n", q2_report);
    println!("=== Team Memo (cloned from memo template) ===");
    println!("{}\n", team_memo);

    // Verify deep copy: original template is unmodified
    let fresh = registry.create("report").unwrap();
    println!("=== Fresh clone (proves original is untouched) ===");
    println!("{}", fresh);
}
